Sources/PackageRegistry/RegistryManager.swift

```diff
@@ -0,0 +1,296 @@
+/*
+ This source file is part of the Swift.org open source project
+
+ Copyright (c) 2014 - 2020 Apple Inc. and the Swift project authors
+ Licensed under Apache License v2.0 with Runtime Library Exception
+
+ See http://swift.org/LICENSE.txt for license information
+ See http://swift.org/CONTRIBUTORS.txt for Swift project authors
+ */
+
+import Basics
+import PackageLoading
+import PackageModel
+
+import TSCBasic
+import TSCUtility
+
+import struct Foundation.URL
+import struct Foundation.URLComponents
+import struct Foundation.URLQueryItem
+
+import Dispatch
+
+public enum RegistryError: Error {
+    case invalidOperation
+    case invalidResponse
+    case invalidURL
+    case invalidChecksum(expected: String, actual: String)
+}
+
+fileprivate let userAgent = "SwiftPackageManager/\(ToolsVersion.currentToolsVersion)"
+
+public final class RegistryManager {
+    internal static var clientFactory: () -> HTTPClientProtocol = {
+        var configuration = HTTPClientConfiguration()
+        configuration.followRedirects = false
+
+        return HTTPClient(configuration: configuration, handler: nil, diagnosticsEngine: nil)
+    }
+
+    private static var cache = ThreadSafeKeyValueStore<URL, RegistryManager>()
+
+    private let registryBaseURL: Foundation.URL
+    private let client: HTTPClientProtocol
+
+    init(registryBaseURL: Foundation.URL) {
+        self.registryBaseURL = registryBaseURL
+        client = Self.clientFactory()
+    }
+
+    public class func discover(
+        for package: PackageReference,
+        completion: @escaping (Result<RegistryManager, Error>) -> Void
+    ) {
+        guard let url = URL(string: "https://\(package.identity)") else {
```

Comment:
this is confusing to me.. are we using the package identity as the URL? what am I missing?

Reply:
Yes, that's correct. This is discussed in [§ 4.1.1 Content Negotiation](https://github.com/mattt/swift-registry-service-proposal/blob/master/package-registry-specification.md#411-content-negotiation). 

The new (non-legacy) package identity is a canonicalized, scheme-agnostic URL. The process of "discovering" registry support involves sending a `HEAD` to that URL with the right headers to see if the server redirects us to a package registry API endpoint.